### ISSUE TITLE
Add Teehistorian

### ISSUE THREAD

**Issue by jxsl13** [teeworlds/teeworlds#2418]
I think the super space efficient and even greatly compressible Teehistorian player input logging might make sense in Vanilla.
Vanilla might be the last to have a problem with bots, but I think if people could have an interchangeable file format that every server uses for logging specific player inputs, the results of such might be usable later on.
Either logging one specific player(everyone while that player is online), a specific IP, or just all players as a kind of space efficient demo method.

Currently heinrich5991's Teehistorian does not yet keep track of entities other than characters, so if I remember correctly you won't be able to see e.g. projectiles in the reconstructed demo.

The general idea is to have a widely used format that might be used for crowdsourcing and maybe  machine learning applications down the line, especially when we come to a point where 0.6 basically ended.

Have fun discussing.

**Comment by Dune-jr** [comment]
Please provide links or doc for those not familiar with that tool.

**Comment by jxsl13** [comment]
It is basically part of DDRace and allows for logging of player inputs.

https://github.com/ddnet/ddnet7/blob/master/src/game/server/teehistorian.h
https://github.com/ddnet/ddnet7/blob/master/src/game/server/teehistorian.cpp

heinrich5991 also provides some conversion tools written in Rust to unpack the format.
https://github.com/heinrich5991/libtw2/tree/master/teehistorian

**Comment by heinrich5991** [comment]
Where do you get from that the teehistorian is in more than one file? (apart from header and source file being two files).

**Comment by heinrich5991** [comment]
> Currently heinrich5991's Teehistorian does not yet keep track of entities other than characters, so if I remember correctly you won't be able to see e.g. projectiles in the reconstructed demo.

This is incorrect. There is no real tool to reconstruct demos. But if we code one, you'll be able to recover the full demo. Teehistorian does contain character position data to check that you're still on track if you try reevaluating the inputs. The `teehistorian2demo` tool uses the character positions and puts them in a demo, therefore it lacks everything but player positions.

**Comment by Dune-jr** [comment]
I'm a bit confused what the use of this would be over demos in vanilla

**Comment by heinrich5991** [comment]
This makes it feasible to store your server's world completely and for a long time. Demos are huge if you record them for a long time.

**Comment by jxsl13** [comment]
> Where do you get from that the teehistorian is in more than one file? (apart from header and source file being two files).

I don't know what exactly you mean.

I kind of pointed out some use cases, even tho teehistorian tracks the whole server, one might want to track a specific player(with specific meta information, basically a reason why that player was tracked) until that player leaves or something like that.

**Comment by jxsl13** [comment]
This should be added :D
